### PR TITLE
.murdock: Fix kconfig hash check

### PR DESCRIPTION
**Pull request by MrKevinWeiss**
### Contribution description

It seems like #18236 introduced some errors in the kconfig hash check.
Running locally I get a `.murdock: line 480: [: too many arguments` message.
On the CI it looks like `./.murdock: 480: [: 4994c097e49decc77a99c2254e46c710238dc4f2: unexpected operator`.
I noticed that the path names are now included in the comparison.
It seems surrounding them in quotes helps but this also compares whitespaces and newlines...

### Testing procedure

Check the murdock output in `tests/shell` for `native` and ensure no errors on line 480.

### Issues/PRs references
Introduced in #18236

### CODE REVIEW

**Review by maribu — APPROVED** [review]

**Comment by MrKevinWeiss** [comment]
Yikes, now we can see all the hash differences.  OK, I guess this will also involve fixes for other things.

File: cpu/esp32/Kconfig.esp32 [hunk 1]
```diff
@@ -24,7 +24,7 @@ config CPU_FAM_ESP32
 
     select PACKAGE_ESP32_SDK if TEST_KCONFIG
 
-    select MODULE_LIBC_GETTIMEOFDAY if TEST_KCONFIG
```
Comment (gschorcht) [inline cpu/esp32/Kconfig.esp32:27]: This line was already removed in PR #18202 with 4dc5e86b2a5e986e206416f9b1e7c26bfb55d882 but readded by mistake when `Kconfig` was splitted into `Kconfig` and `Kconfig.esp32` with PR #18320 as a split-off from PR #17842 :worried:

File: cpu/esp32/Kconfig.esp32 [hunk 2]
```diff
@@ -24,7 +24,7 @@ config CPU_FAM_ESP32
...
+    select MODULE_MALLOC_THREAD_SAFE if !MODULE_ESP_IDF_HEAP && TEST_KCONFIG
```
Comment (gschorcht) [inline cpu/esp32/Kconfig.esp32:27]: This line was already added by PR #18209 commit 041fe58209c51597ec8f6fc45e11c2a6db4b9d93 but removed by mistake when Kconfig was splitted into Kconfig and Kconfig.esp32 with PR #18320 as a split-off from PR #17842

**Review by gschorcht — APPROVED** [review]

File: cpu/esp32/Kconfig.esp32 (hunk 2 above)
Comment (gschorcht) [reply cpu/esp32/Kconfig.esp32:27]: @MrKevinWeiss Good catch, thanks for figuring it out.

**Comment by MrKevinWeiss** [comment]
I hope now it should be OK.  Thanks for looking at it even before the ping @gschorcht 

Wasn't too bad :smile: to find

File: cpu/esp32/esp-idf/gpio/Kconfig [hunk 3]
```diff
@@ -14,3 +14,7 @@ config MODULE_ESP_IDF_GPIO
 
     help
         ESP-IDF code for peripheral GPIO.
+
+config MODULE_ESP_IDF_GPIO_HAL
+    bool "ESP IDF GPIO Hardware Abstraction"
```
Comment (gschorcht) [inline cpu/esp32/esp-idf/gpio/Kconfig:19]:
```suggestion
    bool "Use ESP-IDF GPIO Hardware Abstraction Layer"
```

Comment (MrKevinWeiss) [reply cpu/esp32/esp-idf/gpio/Kconfig]: Done!

**Comment by MrKevinWeiss** [comment]
HAH: I found [where the hash issue is](https://ci.riot-os.org/RIOT-OS/RIOT/18364/43cc07f050ba20e3c889ac871abd5ad76467cefa/output/compile/tests/driver_pcf857x/esp32-heltec-lora32-v2:gnu.txt)

Turns out that the `build/esp_bootloader/bootloader.bin` is responsible.  I really want this in ASAP but we can at least see that #18236 helps show the issue!

**Comment by MrKevinWeiss** [comment]
Can we just ignore the bootloader.bin as it isn't really RIOT based and not affected by Kconfig?

**Comment by MrKevinWeiss** [comment]
I mean ideally get the hashes to match but if that takes some time or is out of our control then...

**Comment by leandrolanzieri** [comment]
Nice catch :)

**Comment by gschorcht** [comment]
> HAH: I found [where the hash issue is](https://ci.riot-os.org/RIOT-OS/RIOT/18364/43cc07f050ba20e3c889ac871abd5ad76467cefa/output/compile/tests/driver_pcf857x/esp32-heltec-lora32-v2:gnu.txt)

The result with this fix seems to be exactly the same again as before. The hash test fails for exactly one application and exactly one board, but each time for a different application and a different board, and even for different architectures and not only the ESP.

I bet that the next time the compilation is run in CI, a different application and board and maybe even a different architecture will fail.

**Comment by MrKevinWeiss** [comment]
> I bet that the next time the compilation is run in CI, a different application and board and maybe even a different architecture will fail.

I would take that bet.  I really think it is the bootloader.  Each unexplainable failure has been ESPs.  Also the fact it is the bootloader binary should be a hint.  I wonder if there is just some timestamp or something in the binary.  I really don't know that world though.  I can try to add a hack to prevent it but maybe it would be better to leave it in to collect more information.

**Comment by MrKevinWeiss** [comment]
> Each unexplainable failure has been ESPs

In my experience... I could be wrong.

**Comment by gschorcht** [comment]
> I would take that bet. I really think it is the bootloader. Each unexplainable failure has been ESPs.

Unfortunatly not, two examples where we had such hash mismatches:
- [tests/driver_ads101x/esp8266-esp-12x:gnu](https://ci.riot-os.org/RIOT-OS/RIOT/18134/7755431ad528eb8e44add12f2c845a8787b8cacb/output/compile/tests/driver_ads101x/esp8266-esp-12x:gnu.txt) which doesn't use the ESP bootloader
- [tests/pkg_utensor/avsextrem:gnu](https://ci.riot-os.org/RIOT-OS/RIOT/17040/1be5cd1061d600593dab6fd49f7dbb763e97b049/output/compile/tests/pkg_utensor/avsextrem:gnu.txt)

**Comment by MrKevinWeiss** [comment]
Hmmm interesting.  Maybe I lost the bet.  They have some noticeable binary size differences though.  I wonder if a rerun would yield the same results.

**Comment by gschorcht** [comment]
> Turns out that the `build/esp_bootloader/bootloader.bin` is responsible.

We had indeed a hash problem with `esp_bootloader` in the past because the compilation time was compiled in for log outputs. But this code is patched by [pkg/esp32_sdk/patches/patches/0024-bootloader-remove-compile-from-banner.patch](https://github.com/RIOT-OS/RIOT/blob/9e283b967d20337e50ad5f7ef32de08f1d8e247f/pkg/esp32_sdk/patches/0024-bootloader-remove-compile-from-banner.patch) and there shouldn't be any difference in code for both compilation runs.

**Comment by gschorcht** [comment]
> Can we just ignore the bootloader.bin as it isn't really RIOT based and not affected by Kconfig?

I think so, because right now there is nothing to configure with RIOT's Kconfig. But that could change in the future.

**Comment by MrKevinWeiss** [comment]
[tests/pkg_utensor/avsextrem:gnu](https://ci.riot-os.org/RIOT-OS/RIOT/17040/1be5cd1061d600593dab6fd49f7dbb763e97b049/output/compile/tests/pkg_utensor/avsextrem:gnu.txt) is not reproducible locally.  I pushed a fixup and I guess we will see tomorrow when everything is done.  Maybe you have a better way.  Maybe we just leave it as is and rerun.  We have a good 6 hours before we have to decide.

File: .murdock [hunk 4]
```diff
@@ -457,7 +457,7 @@ compile() {
                         make -C${appdir} all test-input-hash -j${JOBS:-4}
         RES=$?
         if [ $RES -eq 0 ]; then
-            kconfig_hashes="$(cat ${BINDIR}/test-input-hash.sha1)"
+            kconfig_hashes="$(awk '!/build\/esp_bootloader\/bootloader.bin/' ${BINDIR}/test-input-hash.sha1 )"
```
[the remark this answers is not in the document]
Comment (MrKevinWeiss) [reply .murdock:460]: This only should affect the CI.  This way locally we can still get the hashes.  Though, as you pointed out, this won't really solve the problem.

Shall I implement the suggested change or drop the change completely?

**Comment by MrKevinWeiss** [comment]
I dropped the `awk '!/build\/esp_bootloader\/bootloader.bin/' ${BINDIR}/test-input-hash.sha1` change and hope that it passes.  We can make another PR to try to fix the hash failures but should get this in for now.

**Comment by gschorcht** [comment]
This time Murdock was happy, so let's go with it.

**Comment by gschorcht** [comment]
@MrKevinWeiss Here is the proof that `esp_bootloader` isn't the problem [log for PR #17442](https://ci.riot-os.org/RIOT-OS/RIOT/17442/e6f5fc606e9a787a357183b9fc64c68497420d13/output/compile/tests/pkg_cryptoauthlib_internal-tests/nucleo-g474re:gnu.txt) :wink:

**Comment by gschorcht** [comment]
> @MrKevinWeiss Here is the proof that `esp_bootloader` isn't the problem [[log for PR #17442](https://ci.riot-os.org/RIOT-OS/RIOT/17442/e6f5fc606e9a787a357183b9fc64c68497420d13/output/compile/tests/pkg_cryptoauthlib_internal-tests/nucleo-g474re:gnu.txt)] wink

In this case, it seems that the binaries really differ in code size. Since the failed compilation of the test application is completely unrelated to this PR, but the test application could obviously be compiled successfully in other PRs, the question is why different code sizes occurred in this case.

**Comment by MrKevinWeiss** [comment]
I guess we need to just keep adding diagnostics...  I guess you win the bet :beers:

**Comment by gschorcht** [comment]
> I guess we need to just keep adding diagnostics... I guess you win the bet beers

And again another board and architecture with [hash mismatches](https://ci.riot-os.org/RIOT-OS/RIOT/18380/cc55a4475dc0baf7e8be4baaba9e317396425ae1/output/compile/tests/pkg_cryptoauthlib_internal-tests/remote-revb:gnu.txt) :wink:

**Comment by MrKevinWeiss** [comment]
Hmmm I wonder if it is kconfig or not.  Maybe I would open a PR and rerun a bunch of times comparing the make outputs to see if there are any mismatches with just the make.

**Comment by gschorcht** [comment]
It seems that the hash mismatches occur mainly with `tests/pkg_cryptoauthlib_internal-tests` and only sometimes with other but different applications. Could the problem possibly depend on the compilation time required. Or could it be that in very special cases the hash values of the previous compilation are used as the first hash value.

**Comment by gschorcht** [comment]
Hm, after some days without hash mismatches, last night it happened 4 times for PRs https://github.com/RIOT-OS/RIOT/pull/18381, https://github.com/RIOT-OS/RIOT/pull/18388 and https://github.com/RIOT-OS/RIOT/pull/18389, all for `tests/pkg_cryptoauthlib_internal-tests`: [log](https://ci.riot-os.org/RIOT-OS/RIOT/18381/5a850725045b81d4951d9c045f2ca631c9e01ea0/output/compile/tests/pkg_cryptoauthlib_internal-tests/adafruit-itsybitsy-m4:gnu.txt) [log](https://ci.riot-os.org/RIOT-OS/RIOT/18381/5a850725045b81d4951d9c045f2ca631c9e01ea0/output/compile/tests/pkg_cryptoauthlib_internal-tests/nucleo-l552ze-q:gnu.txt) [log](https://ci.riot-os.org/RIOT-OS/RIOT/18388/86dc086379a3d53373dce0b9b684dbf12922f2e4/output/compile/tests/pkg_cryptoauthlib_internal-tests/dwm1001:gnu.txt) [log](https://ci.riot-os.org/RIOT-OS/RIOT/18389/863a1c2ac155c53ab3338629a8510043a0b63818/output/compile/tests/pkg_cryptoauthlib_internal-tests/nucleo-f103rb:gnu.txt)

**Comment by gschorcht** [comment]
Would it be helpful to additionally print the last modification time of the hash file and the files used for the hash calculation if the hashes differ?

**Comment by MrKevinWeiss** [comment]
I wanted to prepare a PR for the weekend that would add a whole bunch of debugging and reruns.  I would wait for the weekend though as it should use quite a bit of murdock time.  I don't think we need so much merged into master though.  Ideally we get it reproducible then have a fix, making any debug stuff unneeded.

**Comment by gschorcht** [comment]
I don't know whether we should really wait with further debugging until weekend. It seems that most compilations in CI fail at the moment due to the hash mismatch problem. Restarting them again and again uses also a lot of murdock time.

**Pull request merged**